kostal/__init__.py

```diff
@@ -18,88 +18,182 @@
-        'l2_watt': 81,
-        'l3_watt': 105
+        'actot_w': 9,
+        'yield_tot_kwh': 12,
+        'yield_day_wh': 21,
+        'operation_status': 27,
+        'dc1_v': 45,
+        'dc2_v': 69,
+        'dc3_v': 93,
+        'dc1_a': 54,
+        'dc2_a': 78,
+        'dc3_a': 102,
+        'ac1_v': 48,
+        'ac2_v': 72,
+        'ac3_v': 96,
+        'ac1_w': 57,
+        'ac2_w': 81,
+        'ac3_w': 105
     }
 
-    def __init__(self, smarthome, ip, user="pvserver", passwd="pvwr", cycle=300):
-        self._sh = smarthome
-        self.ip = ip
+    def __init__(self, sh, ip, user="pvserver", passwd="pvwr",cycle=300, datastucture="html"):
+        self._sh = sh
+        self.logger = logging.getLogger(__name__)
+        self.logger.info('Init Kostal plugin')
         self.user = user
         self.passwd = passwd
         self.cycle = int(cycle)
-        self._items = []
-        self._values = {}
+        self._items = {}
+        if Utils.is_ip(ip):
+            self.ip = ip
+        else:
+            self.logger.error(str(ip) + " is not a valid IP")
+        if datastucture == "html":
+            self._keytable = self._key2td
+            self.datastucture = "html"
+        else:
+            self._keytable = self._key2json
+            self.datastucture = "json"
 
     def run(self):
+        """
+        Run method for the plugin
+        """
+        self.logger.debug("run method Kostal called")
         self.alive = True
         self._sh.scheduler.add('Kostal', self._refresh, cycle=self.cycle)
 
     def stop(self):
+        """
+        Stop method for the plugin
+        """
+        self.logger.debug("stop method Kostal called")
         self.alive = False
 
     def parse_item(self, item):
-        if 'kostal' in item.conf:
-            kostal_key = item.conf['kostal']
-            if kostal_key in self._key2td:
-                self._items.append([item, kostal_key])
-                return self.update_item
-            else:
-                logger.warn('invalid key {0} configured', kostal_key)
-        return None
-
+        """
+        Default plugin parse_item method. Is called when the plugin is initialized.
+        Selects each item corresponding to its attribute keywords and adds it to an internal array
+        :param item: The item to process.
+        """
+        if self.has_iattr(item.conf, 'kostal'):
+            self._items[self.get_iattr_value(item.conf, 'kostal')] = item
+            self.logger.debug("parse item: {0}".format(item))
+            return self.update_item
     def parse_logic(self, logic):
         pass
-
     def update_item(self, item, caller=None, source=None, dest=None):
-        if caller != 'Kostal':
-            pass
-
+        if item():
+            if self.has_iattr(item.conf, 'kostal'):
+                self.logger.debug("update_item ws called with item '{}' from caller '{}', source '{}' and dest '{}'".format(item, caller, source, dest))
+                pass
```

Comment:
You're right. The example is just a template for the plugin, but I don't think that all plugin should implement and keep this method like shown in this example. It should help to understand what all the methods do.

I would implement the method like this:
```
    def update_item(self, item, caller=None, source=None, dest=None):
        """
        Write items values
        :param item: item to be updated towards the plugin
        :param caller: if given it represents the callers name
        :param source: if given it represents the source
        :param dest: if given it represents the dest
        """
        pass
```